pyiron_gpl/pacemaker/pacemaker.py

```diff
@@ -0,0 +1,231 @@
+
+try:
+    from pyace import BBasisConfiguration, ACEBBasisSet
+
+    HAS_PYACE = True
+except ImportError as e:
+    print("Could not import `pyace` package. The package should be installed for proper functionality")
+    HAS_PYACE = False
+
+# set loggers
+loggers = [logging.getLogger(name) for name in logging.root.manager.loggerDict]
+for logger in loggers:
+    logger.setLevel(logging.WARNING)
+
+class PaceMakerJob(GenericJob):
+    def __init__(self, project, job_name):
+        super().__init__(project, job_name)
+        self.__name__ = "PaceMakerJob"
+        self.__version__ = "0.1"
+
+        self.input = GenericParameters(table_name="input")
+        self.input['cutoff'] = 10.
+        self.input['metadata'] = {}
+        self.input['data'] = {}  # data_config
+        self.input['potential'] = {}  # potential_config
+        self.input['fit'] = {}  # fit_config
+        self.input['backend'] = {'evaluator': 'tensorpot'}  # backend_config
+
+        self.structure_data = None
+
+        # self.executable = "pacemaker input.yaml -l log.txt"
+        self._executable = None
+        self._executable_activate()
+
+    def _save_structure_dataframe_pckl_gzip(self, df):
+        df.rename(columns={"number_of_atoms": "NUMBER_OF_ATOMS",
+                           "energy": "energy_corrected",
+                           "atoms": "ase_atoms"}, inplace=True)
+        df["NUMBER_OF_ATOMS"] = df["NUMBER_OF_ATOMS"].astype(int)
+        if "pbc" not in df.columns:
+            df["pbc"] = df["ase_atoms"].map(lambda atoms: np.all(atoms.pbc))
+
+        data_file_name = os.path.join(self.working_directory, "df_fit.pckl.gzip")
+        logging.info("Saving training structures dataframe into {} with pickle protocol = 4, compression = gzip".format(
+            data_file_name))
+        df.to_pickle(data_file_name, compression="gzip", protocol=4)
+        return data_file_name
+
+    def write_input(self):
+        # prepare datafile
+        if self.structure_data is None:
+            raise ValueError(
+                "`structure_data` is none, but should be pd.DataFrame, TrainingContainer or valid pickle.gzip filename")
+        if isinstance(self.structure_data, pd.DataFrame):
+            logging.info("structure_data is pandas.DataFrame")
+            data_file_name = self._save_structure_dataframe_pckl_gzip(self.structure_data)
+            self.input["data"] = {"filename": data_file_name}
+        elif isinstance(self.structure_data, str):  # filename
+            if os.path.isfile(self.structure_data):
+                logging.info("structure_data is valid file path")
+                self.input["data"] = {"filename": self.structure_data}
+            else:
+                raise ValueError("Provided structure_data filename ({}) doesn't exists".format(self.structure_data))
+        elif hasattr(self.structure_data, "get_pandas"):  # duck-typing check for TrainingContainer
+            logging.info("structure_data is TrainingContainer")
```

Comment:
If this is the case, is it not safer to directly test if the instance is a `TrainingContainer`? I admit I'm not super familiar with the context, so I recognize this might be a dumb question.

Reply:
I'm note sure if we settled up with the `TrainingContainer` name, so I wouldprefer (for the moment) to use duck-typing

Reply:
If this is the case then I would double down on my statement that you should directly test if the instance is a training container. If you `from pyiron_foo import TrainingContainer` and `isinstance(self.structure_data, TrainingContainer)` here, then as long as your pyiron dev environment includes both `pyiron_gpl` and `pyiron_foo` even an extremely simple IDE will catch this occurrence during refactor->rename:TrainingContainer. However, if you later rename the *method* `get_pandas` on that object, then it's hit or miss if the developer will see it in the refactoring change list because it will be tucked down in the string comparison section.

I'm assuming `TrainingContainer` already exists somewhere and you mean you might still rename it. If it doesn't even exist yet I'd just replace the entire section of code with a TODO pending the completion of that class.

Reply:
OK, agreed